### PR TITLE
backend pagintation

### LINKED ISSUE #211: adding backend pagination
## Summary

- add backend pagination when requesting data for submissions, users, and clarifications

### Basic example

- res.send(await getBlueStandings(isPractice, 0)) returns the first 25 teams
### Motivation
- we want to speed up how fast requests are made by limiting how much data is being grabbed at one time

### PR DESCRIPTION
**Pull request by MrKevinOConnell**
# Description

- this PR adds pagination to the backend (manually sets limit to 25)
- if a page number is given, it will skip (page# - 1) * 25  values, then begin counting
- if a page number is not given, it will not have a limit

Fixes #211 

## Type of change
- [x] 💡 New feature

# How Has This Been Tested?

- this is the result that occurs when you go to admin/users on the frontend  (with a signed in account) and the page called is 1 (had page size as 10 for easier readability)
Screen Shot 2021-10-05 at 11 37 56 PM

- this is the result that occurs when you go to admin/users on the frontend (with a signed in account) and the page called is 2 (had page size as 10 for easier readability)
Screen Shot 2021-10-05 at 11 38 34 PM

- FYI - this works on mongo, still need to test with dynamo once we set up a staging/prod environment

- [x] Tested individually.

# Checklist:

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [x] My changes do not break any features.

### CODE REVIEW

File: backend/src/api/standings/getStandings.ts [hunk 1]
```diff
@@ -30,7 +30,12 @@ const getBlueStandings = async (isPractice: boolean): Promise<Record<string, any
   let standings = Object.values(await contest.get_users({ role: 'team', division: 'blue' }))
   const submissions = await contest.get_submissions({ division: 'blue' })
 
-  let problemsList = await contest.get_problems({ division: 'blue' }, ['pid', 'division', 'id', 'name', 'practice'])
+  let problemsList = await contest.get_problems(
+    { division: 'blue' },
+    0,
+    ['pid', 'division', 'id', 'name', 'practice'],
+    0
```
Comment (Jl5her) [inline backend/src/api/standings/getStandings.ts:37]: What is this fourth parameter for? `get_problems()` calls for `args`, `page` and `columns` right?

File: backend/src/services/db/database.ts [hunk 2]
```diff
@@ -12,7 +12,7 @@ export default abstract class Database {
     }
   }
 
-  abstract scan(TableName: string, query?: ScanOptions): Promise<Item[]>
+  abstract scan(TableName: string, query?: ScanOptions, page?: number): Promise<Item[]>
```
Comment (Jl5her) [inline backend/src/services/db/database.ts:15]: This should be implemented for `db/dynamo` as well, not just mongo before merging.

File: backend/src/api/standings/getStandings.ts [hunk 3]
```diff
@@ -126,11 +131,11 @@ const getBlueStandings = async (isPractice: boolean): Promise<Record<string, any
...
 // }
 
-const getGoldStandings = async (isPractice: boolean): Promise<Record<string, any>> => {
+const getGoldStandings = async (isPractice: boolean, page?: number): Promise<Record<string, any>> => {
   let standings = Object.values(await contest.get_users({ role: 'team', division: 'gold' }))
   const submissions = await contest.get_submissions({ division: 'gold' })
-
-  let problemsList = await contest.get_problems({ division: 'gold' }, [
+  //while frontend is being built making page optional
+  let problemsList = await contest.get_problems({ division: 'gold' }, page ? page : undefined, [
```
Comment (Jl5her) [inline backend/src/api/standings/getStandings.ts:138]: Each row of the standings for gold / blue are different teams (users), so pagination would call for reducing the teams displayed, but not until after sorting by score (or solved problems).

Only grabbing a couple of problems (even though there are less than 25 problems right now) can lead to not being able to resolve the `pid` to a problem. 

The submissions can be associated with the universe of problems, but when grabbing only a subset of the universe to cross-reference, it is quite possible there will be conflicts when a problem is in the universe, but not in the subset.

File: backend/src/api/standings/getStandings.ts (hunk 1 above)
Comment (MrKevinOConnell) [reply backend/src/api/standings/getStandings.ts:37]: good catch

File: backend/src/api/standings/getStandings.ts (hunk 3 above)
Comment (MrKevinOConnell) [reply backend/src/api/standings/getStandings.ts:138]: We could just paginate the submissions instead of the problems?

**Pull request closed without merging**